src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -69,10 +69,40 @@ class OpSequencer {
   const ClientRequest* get_last_issued() const {
     return last_issued;
   }
-  void finish_op(const ClientRequest* const op) {
+  void finish_op_in_order(const ClientRequest* const op) {
     assert(op->get_id() > last_completed->get_id());
     last_completed = op;
   }
+  void finish_op_out_of_order(ClientRequest* const this_op) {
+    assert(this_op);
+    for (auto it = OperationRegistryI::op_list::s_iterator_to(*this_op);
+         it != OperationRegistryI::op_list::s_iterator_to(*last_issued);
+         ++it) {
+      // we're iterating over the operation registry of all ClientRequests.
+      // this list aggrates every single instance in the system, and thus
+      // we need to skip operations coming from different client's session
+      // or targeting different PG. The op we're interested in should be
+      // somewhere between the operation that early exists (this_op) and
+      // the most recently registered one.
+      // as the out-of-order exit is supposed to happen on cold paths only,
+      // the extra overhead is a thing we can live with.
+      auto* maybe_next_op = std::addressof(dynamic_cast<ClientRequest&>(*it));
```

Comment:
why `dynamic_cast`? can we just use `static_cast` instead? i am in favor of not using RTTI if we don't have to.

Reply:
Conceptually it acts like an assertion for the assumption I made. Though, not having strong feelings towards it and can change to `static_cast`.

Reply:
@rzarzynski how about using `static_cast<>()`. but also have
```c++
assert(dynamic_cast<ClientRequest&>(*it));
```
?